core/engine/src/object/jsobject.rs

```diff
@@ -620,88 +651,24 @@ Cannot both specify accessors and a value or writable attribute",
     /// Casts to a `BufferObject` if the object is an `ArrayBuffer` or a `SharedArrayBuffer`.
     #[inline]
     pub(crate) fn into_buffer_object(self) -> Result<BufferObject, JsObject> {
-        let obj = self.borrow();
-
-        if obj.is::<ArrayBuffer>() {
-            drop(obj);
+        if self.is::<ArrayBuffer>() {
             // SAFETY: We have verified that the inner data of `self` is of type `ArrayBuffer`.
             return Ok(BufferObject::Buffer(unsafe {
                 self.downcast_unchecked::<ArrayBuffer>()
             }));
         }
```

Comment:
I think we can remove the unsafe here by chaining downcast calls :)

```suggestion
        match self.downcast::<ArrayBuffer>() {
            Ok(buffer) => return Ok(BufferObject::Buffer(buffer)),
            Err(object) => object.downcast::<SharedArrayBuffer>().map(BufferObject::SharedBuffer)
        }
```